### PR TITLE
Use the new broadcast permissions

### PR DESCRIPTION
**Pull request by klssmith**
See the commit messages for full details, but this does a few things:
1. Makes the broadcast endpoints and the links to them work with the existing broadcast permissions and the [new ones that will be added](https://github.com/alphagov/notifications-api/pull/3274)
2. Updates the broadcast user form (and also rewrites some of the roles and permissions code which no longer works with the way we know have our admin roles and database permissions mapped to each other)
3. Adds a paragraph to the broadcast setting form to warn that [all user permissions will be removed when it is submitted](https://github.com/alphagov/notifications-api/pull/3277)

[Pivotal story](https://www.pivotaltracker.com/story/show/178482657)

### Merge after
- [x] https://github.com/alphagov/notifications-api/pull/3274
- [x] https://github.com/alphagov/notifications-api/pull/3277

### Next steps
The plan for deploying this is to remove people's permissions in the database immediately after this is deployed, then to re-add them manually. The routes will work with no changes throughout, but there will be a brief period (between deploying this code and updating the data) where viewing a user's permissions on the "manage users" page will not show the right thing.

### Screenshots for the new options for the single broadcast page
**Looking at a pending broadcast you have created when the service is in trial mode and you only have `create_broadcasts` permission**
scenario_3

**Looking at a pending broadcast someone else has created when the service is in training mode and you only have `create_broadcasts` permission**
scenario_8_training

**Looking at a pending broadcast someone else has created when the service is live and you only have `create_broadcasts` permission**
scenario-8_live

### CODE REVIEW

File: app/models/roles_and_permissions.py [hunk 1]
```diff
@@ -26,19 +22,25 @@
...
     """
     Given a list of database permissions, return a set of roles
 
-    look them up in roles_by_permission, falling back to just passing through from the api if they aren't in the dict
+    A role is returned if all of its database permissions are in the permission list that is passed in.
+    Any permissions in the list that are not database permissions are also returned.
     """
-    return {roles_by_permission.get(permission, permission) for permission in permissions}
+    # TODO: Stop returning unknown database permissions.
+    unknown_database_permissions = {p for p in permissions if p not in all_database_permissions}
```
Comment (klssmith) [inline app/models/roles_and_permissions.py:38]: We rely on this behaviour a lot in the tests, so removing it would involve changing a lot of tests (which use admin roles where they should be using database permissions). I couldn't see it being used anywhere in the "real code" so I'll tidy this up in a later PR (to keep this one from getting too big).

File: app/models/roles_and_permissions.py [hunk 2]
```diff
@@ -26,19 +22,25 @@
 )
 
 broadcast_permissions = (
-    ('send_messages', 'Prepare and approve broadcasts'),
-    ('manage_templates', 'Add and edit templates'),
-    ('manage_service', 'Manage settings and team'),
+    ('create_broadcasts', 'Write alerts and templates'),
```
Comment (klssmith) [inline app/models/roles_and_permissions.py:25]: Could do with some thoughts on the form labels here please @karlchillmaid. 'Write alerts and templates' also lets you cancel or reject an alert (but not approve it).

**Comment by servingUpAces** [comment]
If the user has `create_broadcasts` permission the `Prepare broadcast` button is missing from the `/templates/{template_id}` page.

File: app/models/roles_and_permissions.py (hunk 2 above)
Comment (karlchillmaid) [reply app/models/roles_and_permissions.py:25]: This is probably worth a short pairing session to discuss – mainly because we haven’t reviewed the UI content yet.

We need to consider these labels labels in context with buttons in the UI like: `New alert`, `New template`, `Prepare broadcast` – all of which could change.

Comment (karlchillmaid) [reply app/models/roles_and_permissions.py:25]: FYI @quis 👆

Comment (klssmith) [reply app/models/roles_and_permissions.py:25]: As well as thinking about content for the labels we will also need to think about content for the page to view a single broadcast because there are now two new scenarios that weren't possible before. This is because the `send_messages` permission (which used to let you both create and approve) has now been split into separate permissions for creating and approving. The new scenarios are:
1. I could have a trial mode service and be viewing an alert I created but I cannot approve it (because I only have the `create_broadcasts` permission
2. I am viewing a pending alert someone else made and I can reject it but not approve it. The service could be either live or trial mode.

**Comment by klssmith** [comment]
I've now updated all the links and buttons to show or not show when they should.

File: app/models/roles_and_permissions.py (hunk 2 above)
Comment (karlchillmaid) [reply app/models/roles_and_permissions.py:25]:
```suggestion
    ('create_broadcasts', 'Add new alerts and templates'),
```

Comment (klssmith) [reply app/models/roles_and_permissions.py]: These content changes have been made now

File: app/templates/views/service-settings/service-confirm-broadcast-account-type.html [hunk 3]
```diff
@@ -41,6 +41,10 @@
         </p>
       {% endif %}
 
+      <p class="govuk-body">
+        All team member permissions will be removed.
```
Comment (benthorner) [inline app/templates/views/service-settings/service-confirm-broadcast-account-type.html:45]: How come we're putting this in now, have we changed the form already?

**Review by benthorner — COMMENTED** [review]
Phew, nice work on a very complex change 🎖️.

I like the use of factory function to parametrize the tests, even if some of them are only temporary.

I've got a few more major comments, but they're mostly on the tests - the app code looks almost good to go. I did some playing around of my own with this with various scenarios, and it seems to work as expected.

File: tests/conftest.py [hunk 4]
```diff
@@ -3580,6 +3580,49 @@ def create_active_user_view_permissions(with_unique_id=False):
     )
 
 
+def create_active_user_create_broadcasts_permissions(with_unique_id=False):
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/conftest.py:3599]: 👍 I've updated the three places which were using the user hash directly to use `create_service_one_user` instead

File: app/templates/views/broadcast/view-message.html [hunk 5]
```diff
@@ -79,9 +80,20 @@ <h1 class="govuk-heading-m govuk-!-margin-bottom-3">
...
+        {% elif current_user.has_permissions('create_broadcasts', restrict_admin_usage=True) %}
+          <p class="govuk-body">
+            You need another member of your team to approve this alert.
+          </p>
+          <p class="govuk-body">
+            This service is in training mode. No real alerts will be sent.
+          </p>
+          {{ page_footer(
+            delete_link=url_for('main.reject_broadcast_message', service_id=current_service.id, broadcast_message_id=broadcast_message.id),
+            delete_link_text='Discard this alert'
```
[the remark this answers is not in the document]
Comment (klssmith) [reply app/templates/views/broadcast/view-message.html:92]: I've copied the relevant part of Karl's content doc below - the plan was to use discard for a trial alert that you create. We agreed to use the new language for the new content in this PR, but to update the existing content in a later PR since the diff on this one is already fairly big.

```
There are 4 destructive actions:

Discard – what you can do to your own alert up until it’s approved
Reject – what someone does to someone else’s alert (or to your own if the service is in training mode and have the right permissions)
Stop sending – what anyone does to an alert that’s been approved and is being sent
Delete – what you do to a template
```

File: app/templates/views/service-settings/service-confirm-broadcast-account-type.html (hunk 3 above)
Comment (klssmith) [reply app/templates/views/service-settings/service-confirm-broadcast-account-type.html:45]: No, that change hasn't been merged yet but I am about to merge it soon so the change will have been made by the time this is deployed.

File: tests/app/main/views/test_dashboard.py [hunk 6]
```diff
@@ -1214,7 +1214,7 @@ def test_menu_send_messages(
             notify_admin,
             api_user_active,
             service_one,
-            ['view_activity', 'send_messages'])
+            ['view_activity', 'send_texts', 'send_emails', 'send_letters'])
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_dashboard.py:1217]: Yes, that's right. This change isn't really needed here, though it does make the test data more realistic.

I would like to get rid of the behaviour to handle unknown database permissions later - I had started to make that change here until I realised that it would need big changes to the test data so it would be clearer in a separate PR.

File: app/templates/views/service-settings/service-confirm-broadcast-account-type.html (hunk 3 above)
Comment (benthorner) [reply app/templates/views/service-settings/service-confirm-broadcast-account-type.html:45]: Oops, sorry: I can see you linked to it in the PR description.

And [I'm the one that reviewed it](https://github.com/alphagov/notifications-api/pull/3277#pullrequestreview-690660689). 🙈

File: tests/app/main/views/test_dashboard.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/app/main/views/test_dashboard.py:1217]: Sure, that's fine with me: I just wanted to check that was the reason for the changes 👍.

File: app/templates/views/broadcast/view-message.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/templates/views/broadcast/view-message.html:92]: Thanks for explaining.

 I agree we can defer updating the content, especially as it looks like we want to change more than just "Withdraw".

File: tests/app/main/test_permissions.py [hunk 7]
```diff
@@ -24,6 +24,28 @@ def test_translate_permissions_from_db_to_admin_roles():
...
+    (
+        ['manage_templates', 'create_broadcasts', 'reject_broadcasts', 'cancel_broadcasts'],
+        {'create_broadcasts', 'manage_templates'},
+    ),
+    (
+        ['manage_templates'],
+        {'manage_templates'},
+    )
+])
+def test_translate_permissions_from_db_to_admin_roles_when_db_role_belongs_to_multiple_admin_roles(
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/test_permissions.py]: Good point, I have combined these into a single test

File: tests/app/main/views/test_broadcast.py [hunk 8]
```diff
@@ -2063,13 +2138,17 @@ def test_confirm_approve_non_training_broadcasts_errors_if_not_ticked(
 
 
 @freeze_time('2020-02-22T22:22:22.000000')
+@pytest.mark.parametrize('user', [
+    create_active_user_with_permissions(),
+    create_active_user_approve_broadcasts_permissions(),
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py:2143]: 👍 Now added

File: tests/app/main/views/test_broadcast.py [hunk 9]
```diff
@@ -1129,13 +1165,19 @@ def test_choose_broadcast_area_page(
     ]
 
 
+@pytest.mark.parametrize('user', (
+    create_active_user_with_permissions(),
+    create_active_user_create_broadcasts_permissions(),
+))
 def test_choose_broadcast_area_page_for_area_with_sub_areas(
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py:1189]: I've now used the same parameterization for the `test_choose_broadcast_sub_area_page_for_district_shows_checkboxes_for_wards` test

File: app/templates/views/broadcast/view-message.html [hunk 10]
```diff
@@ -19,9 +18,10 @@
 
 {% block service_page_title %}
   {% if broadcast_message.status == 'pending-approval' %}
-    {% if broadcast_message.created_by and broadcast_message.created_by == current_user and current_user.has_permissions('send_messages', restrict_admin_usage=True) %}
+    {% if broadcast_message.created_by and broadcast_message.created_by == current_user
+      and current_user.has_permissions('send_messages', 'create_broadcasts', 'approve_broadcasts', restrict_admin_usage=True) %}
```
[the remark this answers is not in the document]
Comment (klssmith) [reply app/templates/views/broadcast/view-message.html:22]: I think for this check I agree that it does seem a little unnecessary to have a different page title / header for similar scenarios. We show `This alert is waiting for approval` when someone created an alert and is viewing it with no permissions vs `<template name> is waiting for approval` when they created an alert and do still have permissions. Since that was already the case, I think this might be one to leave for now and check with Karl later.

Where we are using the same conditional for the actual page content, there are more noticeable differences, so I think it is needed. I made a list of all the options for the page [here](https://docs.google.com/spreadsheets/d/1MtUNH8y7-Ef3N6SWj3mX0qpkN_-J0SeCrPKVTocIk38/edit#gid=0) along with [screenshots for the current content](https://docs.google.com/document/d/1cRJmEf8sGysC5widRt9VJzx57YHvOjm13PHahkcLXVo/edit) that we have to try and understand things better. And you can see the new content for the two scenarios we didn't have [here](https://docs.google.com/document/d/1QdbfXSPB22wDdyLQ18Cc_8mhAmS3781W1yZi21bKM9E/edit).

[the remark this answers is not in the document]
Comment (klssmith) [reply app/templates/views/broadcast/view-message.html:22]: I've added a new test, `test_view_only_user_cant_approve_broadcasts_they_created`, since the existing test of `test_view_only_user_cant_approve_broadcast` was not specifically testing the case when the viewer created the alert. There's some repetition in the tests, but I'm planning to go through to try and simplify the tests when the old permissions are removed.

[the remark this answers is not in the document]
Comment (benthorner) [reply app/templates/views/broadcast/view-message.html:22]: Thanks, that looks good to me 👍. I was wondering if we could simplify the tests somehow, but aside from making them a bit more consistent, I agree that can be left for a follow-up PR.

> it does seem a little unnecessary to have a different page title / header for similar scenarios

I agree the minor difference in content seems unnecessary, but also that it's out-of-scope to change for this PR 👍.

> Where we are using the same conditional for the actual page content, there are more noticeable differences

That's true, but that wouldn't be affected by removing the extra conditions (except for the new test you've added!):

```
    {% if broadcast_message.created_by and broadcast_message.created_by == current_user %}
      <div class="banner govuk-!-margin-bottom-6">
        <h1 class="govuk-heading-m govuk-!-margin-bottom-3">
          {{ broadcast_message.template.name }} is waiting for approval
        </h1>
        {% if current_service.live %}
          <p class="govuk-body">
            You need another member of your team to approve your alert.
          </p>
          {%  if current_user.has_permissions('send_messages', 'create_broadcasts', 'approve_broadcasts', restrict_admin_usage=True)
             {{ page_footer(
               delete_link=url_for('main.reject_broadcast_message', service_id=current_service.id, broadcast_message_id=broadcast_message.id),
               delete_link_text='Discard this alert'
             ) }}
          {% endif %}
        {% elif current_user.has_permissions('send_messages', 'approve_broadcasts', restrict_admin_usage=True) %}
...
```

In other words, they'd see the same content, but not the "Discard" button. This conditional block is pretty complex, so ideally we should simplify it if we can, so it's easier to follow.

Still, a minor point - I just wanted to clarify what I meant.

File: tests/app/main/views/test_manage_users.py [hunk 11]
```diff
@@ -1186,16 +1191,12 @@ def test_invite_user_with_email_auth_service(
     (
         {
             'permissions_field': [
-                'send_messages',
-                'manage_templates',
-                'manage_service',
+                'create_broadcasts',
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_manage_users.py:1233]: We have [this test](https://github.com/alphagov/notifications-admin/blob/e72a260e13fc38079d702ee5fb1e831e7b3c357b/tests/app/main/views/test_manage_users.py#L1253) which tests the POST request when inviting a user to a broadcast service. Was the extra test you were thinking of one that tests viewing the invite page to test the form labels show the right thing?

File: app/templates/views/broadcast/view-message.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (klssmith) [reply app/templates/views/broadcast/view-message.html:22]: I see what you mean now, so the heading would change but not the main page content

File: tests/app/main/views/test_broadcast.py [hunk 12]
```diff
@@ -477,13 +479,19 @@ def test_empty_broadcast_dashboard(
 
 
 @freeze_time('2020-02-20 02:20')
+@pytest.mark.parametrize('user', [
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py:480]: 👍 I'll leave this one for the next PR where the parametrization will be removed

File: tests/app/main/views/test_broadcast.py [hunk 13]
```diff
@@ -1921,12 +1931,16 @@ def test_view_pending_broadcast_without_template(
 
 
 @freeze_time('2020-02-22T22:22:22.000000')
+@pytest.mark.parametrize('user', [
+    create_active_user_with_permissions(),
+    create_active_user_approve_broadcasts_permissions(),
+])
 def test_view_pending_broadcast_from_api_call(
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py:1956]: I've added a new commit which attempts to do that

File: app/templates/views/broadcast/view-message.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/templates/views/broadcast/view-message.html:22]: Yeah that's right. Is there a card / task for follow-up content tweaks? 

I'd advocate that we should simplify this logic as part of doing that.

File: tests/app/main/views/test_manage_users.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/app/main/views/test_manage_users.py:1233]: Sort of: we have [this test](https://github.com/alphagov/notifications-admin/blob/e9a05c04fb58b7500823e176a0752ffdde890f40/tests/app/main/views/test_manage_users.py#L198) that checks permissions show up properly for two active users, which is similar (but different) to [this other test](https://github.com/alphagov/notifications-admin/blob/e9a05c04fb58b7500823e176a0752ffdde890f40/tests/app/main/views/test_manage_users.py#L187) which shows the permissions work with invited users (no idea why we need 5 of them!). That said, I can see it's the same code for pending and active users alike, so a very optional addition I think.

> viewing the invite page to test the form labels show the right thing

I can see we've already got this covered with `test_broadcast_service_only_shows_relevant_permissions` 👍.

Is that a bit clearer?

File: app/templates/views/broadcast/view-message.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (klssmith) [reply app/templates/views/broadcast/view-message.html:22]: No card yet, but it's on my to-do list to make the rest of the content changes that have been agreed.

File: tests/app/main/views/test_manage_users.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_manage_users.py:1233]: Yep, thanks 👍

File: tests/app/main/views/test_broadcast.py [hunk 14]
```diff
@@ -623,11 +629,17 @@ def test_rejected_broadcasts_page(
     )
 
 
+@pytest.mark.parametrize('user', (
+    create_active_user_with_permissions(),
+    create_active_user_create_broadcasts_permissions(),
+))
 def test_new_broadcast_page(
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py:653]: I've added parametrization for the `preview_broadcast_areas` test.

Seems like we did already have [this test](https://github.com/alphagov/notifications-admin/blob/6ac593aa5f176fe9dacb4be51ebda818bb9e8e1b/tests/app/main/views/test_broadcast.py#L2203) which tests the approve endpoint, but I had not updated that to also use the new permissions so I've now added parametrization to that test too (just added that to the the last commit since trying to rebase that properly was getting tricky).

**Comment by klssmith** [comment]
@benthorner I think this is ready for another look now

**Review by benthorner — APPROVED** [review]
Thanks for looking into my comments 🏅.

I found a couple of things that are worth fixing in the tests. However, they all predate this PR, so while some of them made it a fair bit harder to review and I think we should fix them, I don't mind if it's a later thing.

File: tests/app/main/views/test_broadcast.py [hunk 15]
```diff
@@ -2367,6 +2451,11 @@ def test_no_view_page_for_draft(
     )
 
 
+@pytest.mark.parametrize('user', (
+    create_active_user_with_permissions(),
+    create_active_user_create_broadcasts_permissions(),
+    create_active_user_approve_broadcasts_permissions(),
+))
 @pytest.mark.parametrize("user_is_platform_admin", [True, False])
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py]: This is now changed (and added to the last commit)

File: tests/app/main/views/test_broadcast.py [hunk 16]
```diff
@@ -2498,6 +2502,7 @@ def test_request_approval(
     fake_uuid,
```
[the remark this answers is not in the document]
Comment (klssmith) [reply tests/app/main/views/test_broadcast.py:2502]: 👍 done

**Comment by klssmith** [comment]
I've made the two small changes (to change a test name and parametrize the platform admin user). These tests are all going to need to be revisited very soon when we remove the old-style permissions, so I'll leave the suggested changes to the approve tests until then, since that's a slightly bigger change.
